File: frontend/src/api/makV2Observer.ts
import { request } from './client'

export interface MakV2EnablePayload {
  manual_approval: boolean
  testnet_mode: boolean
  reason: string
  confirm_testnet_only: boolean
  confirm_no_production: boolean
  confirm_max_one_lot: boolean
  confirm_no_auto_promotion: boolean
}

export interface MakV2DisablePayload {
  reason: string
}

export interface MakV2DryRunSignalPayload {
  instrument: 'lc' | 'ps'
  exact_contract: string
  side: 'long' | 'short'
  z_score: number
  rolling_mean?: number | null
  rolling_std?: number | null
  last_price: number
  bid_price_1: number
  ask_price_1: number
  bid_volume_1: number
  ask_volume_1: number
  quote_age_ms: number
  cluster_id: string
  active_overlap_900s: number
  cooldown_state: string
  data_quality_status: 'pass' | 'degraded' | 'blocked'
}

export interface MakV2ObserverStatus {
  mode: string
  candidate_id: string
  profile_id: string
  capacity_status: string
  enabled: boolean
  manual_approval: boolean
  testnet_mode: boolean
  dry_run_only: boolean
  production_allowed: boolean
  max_order_lots: number
  max_testnet_orders_per_day: number
  signals_total: number
  dry_run_intents_total: number
  blocked_signals_total: number
  guardrail_events_total: number
  order_endpoint_touched: boolean
  enable_rejected?: boolean
}

export interface MakV2SafetyAuditPayload {
  probe_rpc: boolean
  collect_rpc_snapshot: boolean
  require_rpc_connected: boolean
  expected_exact_contracts: string[]
}

export interface MakV2SafetyAuditCheck {
  name: string
  status: 'PASS' | 'WATCH' | 'FAIL'
  observed: unknown
}

export interface MakV2SafetyAuditResult {
  audit_time_utc: string
  mode: string
  overall: 'PASS' | 'WATCH' | 'FAIL'
  single_order_smoke_allowed: boolean
  checks: MakV2SafetyAuditCheck[]
  observer: Record<string, unknown>
  risk: Record<string, unknown>
  trade_config: Record<string, unknown>
  rpc: Record<string, unknown>
  snapshot: {
    accounts: Record<string, unknown>[]
    mak_positions: Record<string, unknown>[]
    gfex_contracts: Record<string, unknown>[]
    errors: Record<string, unknown>
  }
  next_actions: string[]
}

export type MakV2SafetyAuditLatest = Partial<MakV2SafetyAuditResult>

export const getMakV2Status = () => request<MakV2ObserverStatus>('/api/mak-v2/testnet-observer/status')
export const getMakV2Signals = () => request<Record<string, unknown>[]>('/api/mak-v2/testnet-observer/signals')
export const getMakV2Orders = () => request<Record<string, unknown>[]>('/api/mak-v2/testnet-observer/orders')
export const getMakV2Fills = () => request<Record<string, unknown>[]>('/api/mak-v2/testnet-observer/fills')
export const getMakV2DailySummary = () => request<Record<string, unknown>[]>('/api/mak-v2/testnet-observer/daily-summary')
export const getMakV2Guardrails = () => request<Record<string, unknown>[]>('/api/mak-v2/testnet-observer/guardrails')
export const getMakV2SafetyAuditLatest = () =>
  request<MakV2SafetyAuditLatest>('/api/mak-v2/testnet-observer/safety-audit/latest')
export const listMakV2SafetyAudits = (limit = 50) =>
  request<MakV2SafetyAuditResult[]>(`/api/mak-v2/testnet-observer/safety-audits?limit=${encodeURIComponent(String(limit))}`)

export const enableMakV2Observer = (payload: MakV2EnablePayload) =>
  request<MakV2ObserverStatus>('/api/mak-v2/testnet-observer/enable', { method: 'POST', body: JSON.stringify(payload) })

export const disableMakV2Observer = (payload: MakV2DisablePayload) =>
  request<MakV2ObserverStatus>('/api/mak-v2/testnet-observer/disable', { method: 'POST', body: JSON.stringify(payload) })

export const flattenMakV2Testnet = () =>
  request<Record<string, unknown>>('/api/mak-v2/testnet-observer/flatten-testnet', { method: 'POST' })

export const dryRunMakV2Signal = (payload: MakV2DryRunSignalPayload) =>
  request<Record<string, unknown>>('/api/mak-v2/testnet-observer/dry-run/signal', { method: 'POST', body: JSON.stringify(payload) })

export const runMakV2SafetyAudit = (payload: MakV2SafetyAuditPayload) =>
  request<MakV2SafetyAuditResult>('/api/mak-v2/testnet-observer/safety-audit', { method: 'POST', body: JSON.stringify(payload) })

File: backend/app/schemas/mak_v2_observer.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


MakV2Instrument = Literal["lc", "ps"]
MakV2Side = Literal["long", "short"]
MakV2DataQuality = Literal["pass", "degraded", "blocked"]


class MakV2ObserverEnableRequestDTO(BaseModel):
    manual_approval: bool = False
    testnet_mode: bool = False
    reason: str = Field(min_length=8, max_length=240)
    confirm_testnet_only: bool = False
    confirm_no_production: bool = False
    confirm_max_one_lot: bool = False
    confirm_no_auto_promotion: bool = False


class MakV2ObserverDisableRequestDTO(BaseModel):
    reason: str = Field(default="manual disable", min_length=4, max_length=240)


class MakV2DryRunSignalRequestDTO(BaseModel):
    instrument: MakV2Instrument
    exact_contract: str = Field(min_length=4, max_length=32)
    signal_time_utc: Optional[datetime] = None
    side: MakV2Side
    z_score: float
    rolling_mean: Optional[float] = None
    rolling_std: Optional[float] = None
    last_price: float = Field(gt=0)
    bid_price_1: float = Field(gt=0)
    ask_price_1: float = Field(gt=0)
    bid_volume_1: float = Field(ge=0)
    ask_volume_1: float = Field(ge=0)
    quote_age_ms: int = Field(default=0, ge=0)
    cluster_id: str = Field(default="manual_dry_run", max_length=80)
    active_overlap_900s: int = Field(default=0, ge=0)
    cooldown_state: str = Field(default="clear", max_length=80)
    data_quality_status: MakV2DataQuality = "pass"


class MakV2ObserverStatusDTO(BaseModel):
    mode: str
    candidate_id: str
    profile_id: str
    capacity_status: str
    enabled: bool
    manual_approval: bool
    testnet_mode: bool
    dry_run_only: bool
    production_allowed: bool
    max_order_lots: int
    max_testnet_orders_per_day: int
    max_testnet_orders_per_instrument_per_day: int
    max_active_testnet_positions_total: int
    max_active_testnet_position_per_instrument: int
    cooldown_after_testnet_order_seconds: int
    cooldown_after_reject_seconds: int
    signals_total: int
    dry_run_intents_total: int
    blocked_signals_total: int
    guardrail_events_total: int
    order_endpoint_touched: bool
